src/XMakeBuildEngine/Definition/ToolsetReader.cs

```diff
@@ -471,6 +486,113 @@ bool accumulateProperties
+                var v35Dir = FrameworkLocationHelper.GetPathToDotNetFrameworkV35(DotNetFrameworkArchitecture.Current)
+                             + Path.DirectorySeparatorChar;
+
+                if (root == null)
+                {
+                    var libraryPath = NativeMethodsShared.FrameworkBasePath;
+                    if (toolsPath.StartsWith(libraryPath))
+                    {
+                        root = Path.GetDirectoryName(toolsPath);
+                        if (toolsPath.EndsWith("bin"))
+                        {
+                            root = Path.GetDirectoryName(root);
+                        }
+                    }
+                    else
+                    {
+                        root = libraryPath;
+                    }
+                }
+
+                root += Path.DirectorySeparatorChar;
+
+                // Global properties cannot be overwritten
+                if (globalProperties["FrameworkSDKRoot"] == null && properties["FrameworkSDKRoot"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("FrameworkSDKRoot", root, true, false));
+                }
+                if (globalProperties["MSBuildToolsRoot"] == null && properties["MSBuildToolsRoot"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildToolsRoot", root, true, false));
+                }
+                if (globalProperties["FrameworkSDKRoot"] == null && properties["FrameworkSDKRoot"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("FrameworkSDKRoot", root, true, false));
+                }
+                if (globalProperties["MSBuildFrameworkToolsPath"] == null
+                    && properties["MSBuildFrameworkToolsPath"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildFrameworkToolsPath", toolsPath, true, false));
+                }
+                if (globalProperties["MSBuildFrameworkToolsPath32"] == null
+                    && properties["MSBuildFrameworkToolsPath32"] == null)
+                {
+                    properties.Set(
+                        ProjectPropertyInstance.Create("MSBuildFrameworkToolsPath32", toolsPath, true, false));
+                }
+                /*if (globalProperties["MSBuildToolsPath"] == null && properties["MSBuildToolsPath"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildToolsPath", toolsPath, true, false));
+                }
+                if (globalProperties["MSBuildToolsPath32"] == null && properties["MSBuildToolsPath32"] == null)
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildToolsPath32", toolsPath, true, false));
+                }*/
+                if (globalProperties["MSBuildRuntimeVersion"] == null && properties["MSBuildRuntimeVersion"] == null)
```

Comment:
FWIW I believe MSBuildRuntimeVersion is 4.0.30319 for all modern (4.0+) toolsets.